5. Unspecified behavior for red

red . + does something given inputs of size 3. What is supposed to
happen? Document and test for it.

----------------------------------------------------------------------

10. sort all by itself produces None as output

This seems wrong:

     M jao@cheese:~/git/marcel$ sort
     None

----------------------------------------------------------------------

11. ls () works?!

M jao@cheese:~/git/marcel$ ls ()
drwxr-xr-x   jao      jao              4096   2020 Feb 25 15:33:00   .
drwxr-xr-x   jao      jao              4096   2020 Feb 26 23:43:50   .git
-rw-r--r--   jao      jao                20   2020 Jan 24 17:07:26   .gitignore
drwxr-xr-x   jao      jao              4096   2020 Feb 26 23:59:32   .idea
-rw-r--r--   jao      jao              6420   2020 Feb 25 15:59:14   README.md
drwxr-xr-x   jao      jao              4096   2020 Feb 25 11:22:15   bin
drwxr-xr-x   jao      jao              4096   2020 Feb 21 13:22:22   experiments
drwxr-xr-x   jao      jao              4096   2020 Feb 26 09:35:03   notes
drwxr-xr-x   jao      jao              4096   2020 Feb 19 10:31:26   src
drwxr-xr-x   jao      jao              4096   2020 Feb 25 09:38:13   test

----------------------------------------------------------------------

23. Space after tab completion is off

Want a space when the choice is unambiguous. Don't want it otherwise,
e.g. complete xy with strings xya xyb.

----------------------------------------------------------------------

24. cat problem causes marcel to crash

M jao@cheese:/tmp$ cat /tmp/nsm*
Traceback (most recent call last):
  File "/home/jao/git/marcel/src/marcel/main.py", line 122, in <module>
    MAIN.run()
  File "/home/jao/git/marcel/src/marcel/main.py", line 77, in run
    self.run_command(line)
  File "/home/jao/git/marcel/src/marcel/main.py", line 89, in run_command
    Command(pipeline).execute()
  File "/home/jao/git/marcel/src/marcel/main.py", line 40, in execute
    self.pipeline.receive_complete()
  File "/home/jao/git/marcel/src/marcel/core.py", line 240, in receive_complete
    self.first_op.receive_complete()
  File "/home/jao/git/marcel/src/marcel/op/bash.py", line 61, in receive_complete
    self.runner.run()
  File "/home/jao/git/marcel/src/marcel/op/bash.py", line 85, in run
    stdout, stderr = process.communicate(input=input)
  File "/usr/lib/python3.7/subprocess.py", line 964, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.7/subprocess.py", line 1755, in _communicate
    self.stdout.errors)
  File "/usr/lib/python3.7/subprocess.py", line 841, in _translate_newlines
    data = data.decode(encoding, errors)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 1247: invalid start byte


Also:

M jao@cheese:~$ find teaching
Process Process-3:
Traceback (most recent call last):
  File "/usr/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/jao/git/marcel/src/marcel/job.py", line 130, in run_command
    env_vars = command.execute()
  File "/home/jao/git/marcel/src/marcel/main.py", line 43, in execute
    self.pipeline.receive_complete()
  File "/home/jao/git/marcel/src/marcel/core.py", line 300, in receive_complete
    self.first_op.receive_complete()
  File "/home/jao/git/marcel/src/marcel/op/bash.py", line 79, in receive_complete
    self.runner.run()
  File "/home/jao/git/marcel/src/marcel/op/bash.py", line 124, in run
    stdout, stderr = process.communicate(input=input)
  File "/usr/lib/python3.7/subprocess.py", line 964, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.7/subprocess.py", line 1755, in _communicate
    self.stdout.errors)
  File "/usr/lib/python3.7/subprocess.py", line 841, in _translate_newlines
    data = data.decode(encoding, errors)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x87 in position 1113486: invalid start byte


----------------------------------------------------------------------

34. Tab completion vs. sudo -- shell crash

If we enter this much of a command:

    sudo -i [ ls /home/jao/.dbus/ses

tab-completion runs into permission problems trying to do its work.

----------------------------------------------------------------------

47. Emacs in background doesn't work

- Run emacs
- Ctrl-Z

Goes to background, but bg and fg do not revive it.

Complete "help interaction" when this is working.

----------------------------------------------------------------------

48. Syntax error in .marcel.py should be handled more gracefully

jao@cheese:~/git/marcel/test$ marcel
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/jao/git/marcel/marcel/main.py", line 211, in <module>
    MAIN = Main()
  File "/home/jao/git/marcel/marcel/main.py", line 90, in __init__
    self.env = marcel.env.Environment(self.config)
  File "/home/jao/git/marcel/marcel/env.py", line 136, in __init__
    self.read_config(config_file)
  File "/home/jao/git/marcel/marcel/env.py", line 201, in read_config
    exec(config_source, self.namespace, locals)
  File "<string>", line 64
    cat = [ map (f: (f, f.readlines())) | expand 1 ]
                  ^
SyntaxError: invalid syntax
----------------------------------------------------------------------

52. Pickling/unpickling Environment

What is the right thing to do? Parts of it make no sense for farcel:
color scheme, PWD, probably others. HOMEDIR may or may not be right.
Also, sometimes we copy for remote execution, sometimes for local
execution. 

Need to rethink Environment transmission.

----------------------------------------------------------------------

58. Infinite loop in help

There was a typo in the HELP for ps:

HELP = '''
{L,wrap=F}ps [-u|--user [USER]] [-g|--group [GROUP}] [-p|--pid PID] [-c|--command STRING]

Note that GROUP is followed by } instead of ]. That seemed to put 
help into an infinite loop.

----------------------------------------------------------------------

59. cat can't find file that exists

Is bash not expanding ~?

M-0.7.12 jao@cheese:~/git/marcel/test$ cat ~/system76.txt
Error(cat: '~/system76.txt': No such file or directory)

----------------------------------------------------------------------

60. Preserve quotes when passing args to bash and sudo

Quote-sensitive tests in test_ops fail otherwise.

RE-ENABLE test_bash() in test_ops once this is fixed.

